matlab/src/feather_writer.cc

```diff
@@ -0,0 +1,365 @@
+  // - However arrow::Buffer must not free this region by itself, since that could cause
+  //   segfaults if the input array is used later in MATLAB.
+  //   - The Doxygen doc for arrow::Buffer's constructor implies that it is not an RAII
+  //     type, so this should be safe from possible double-free here.
+  std::shared_ptr<arrow::Buffer> buffer(
+      new arrow::Buffer(reinterpret_cast<const uint8_t*>(dt),
+                        mxGetElementSize(data) * mxGetNumberOfElements(data)));
+
+  // Construct arrow::NumericArray specialization using arrow::Buffer.
+  // Pass in nulls information...we could compute and provide the number of nulls here too,
+  // but passing -1 for now so that Arrow recomputes it if necessary.
+  std::unique_ptr<arrow::Array> array_wrapper(new arrow::NumericArray<ArrowDataType>(
+      mxGetNumberOfElements(data), buffer, validity_bitmap, -1));
+
+  return array_wrapper;
+}
+
+// Dispatch MATLAB column data to the correct arrow::Array converter.
+std::unique_ptr<arrow::Array> WriteVariableData(
+    const mxArray* data, const std::string& type,
+    const std::shared_ptr<arrow::Buffer> validity_bitmap) {
+  // Get the underlying type of the mxArray data.
+  const mxClassID mxclass = mxGetClassID(data);
+
+  switch (mxclass) {
+    case mxSINGLE_CLASS:
+      return WriteNumericData<arrow::FloatType>(data, validity_bitmap);
+    case mxDOUBLE_CLASS:
+      return WriteNumericData<arrow::DoubleType>(data, validity_bitmap);
+    case mxUINT8_CLASS:
+      return WriteNumericData<arrow::UInt8Type>(data, validity_bitmap);
+    case mxUINT16_CLASS:
+      return WriteNumericData<arrow::UInt16Type>(data, validity_bitmap);
+    case mxUINT32_CLASS:
+      return WriteNumericData<arrow::UInt32Type>(data, validity_bitmap);
+    case mxUINT64_CLASS:
+      return WriteNumericData<arrow::UInt64Type>(data, validity_bitmap);
+    case mxINT8_CLASS:
+      return WriteNumericData<arrow::Int8Type>(data, validity_bitmap);
+    case mxINT16_CLASS:
+      return WriteNumericData<arrow::Int16Type>(data, validity_bitmap);
+    case mxINT32_CLASS:
+      return WriteNumericData<arrow::Int32Type>(data, validity_bitmap);
+    case mxINT64_CLASS:
+      return WriteNumericData<arrow::Int64Type>(data, validity_bitmap);
+
+    default: {
+      mexErrMsgIdAndTxt("MATLAB:arrow:UnsupportedArrowType",
+                        "Unsupported arrow::Type '%s' for variable '%s'",
+                        mxGetClassName(data), type.c_str());
+    }
+  }
+
+  // We shouldn't ever reach this branch, but if we do, return a null unique_ptr.
+  return std::unique_ptr<arrow::Array>();
+}
+
+}  // namespace internal
+
+arrow::Status FeatherWriter::Open(const std::string& filename,
+                                  std::shared_ptr<FeatherWriter>* feather_writer) {
+  // Allocate shared_ptr out parameter.
+  *feather_writer = std::shared_ptr<FeatherWriter>(new FeatherWriter());
```

Comment:
`make_shared<...>(...)`?

Reply:
Currently, the default constructor for `FeatherWriter` is marked `private` to prevent clients from directly instantiating instances of `FeatherWriter` without calling `FeatherWriter::Open`. `std::make_shared` requires access to a `public` constructor for the type it is allocating memory for.

@wesm Do you think it is better to maintain the use of `new` here and keep the `FeatherWriter` constructor `private` or switch to a `public` constructor and use `std::make_shared` with the risk of clients instantiating invalid `FeatherWriter` objects (e.g. if `feather::ipc::TableWriter::Open` fails)?

Reply:
Great, thanks for clearing that up. I'll stick with `new` in that case.